askama/src/filters/json.rs

```diff
@@ -17,10 +18,58 @@ use serde_json::to_writer_pretty;
 /// To use it in HTML attributes, you can either use it in quotation marks `"{{data|json}}"` as is,
 /// or in apostrophes with the (optional) safe filter `'{{data|json|safe}}'`.
 /// In HTML texts the output of e.g. `<pre>{{data|json|safe}}</pre>` is safe, too.
-pub fn json<S: Serialize>(s: S) -> Result<String> {
-    let mut writer = JsonEscapeBuffer::new();
-    to_writer_pretty(&mut writer, &s).map_err(Error::from)?;
-    Ok(writer.finish())
+#[inline]
+pub fn json<S: Serialize>(s: S) -> Result<ToJson<S>, Infallible> {
+    Ok(ToJson(s))
+}
+
+#[derive(Debug, Clone)]
+pub struct ToJson<S: Serialize>(S);
+
+impl<S: Serialize> fmt::Display for ToJson<S> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        to_writer_pretty(JsonWriter(f), &self.0).map_err(|_| fmt::Error)
```

Comment:
```suggestion
        #[cfg(debug_assertions)]
        to_writer_pretty(JsonWriter(f), &self.0).map_err(|_| fmt::Error)
        #[cfg(not(debug_assertions))]
        to_writer(JsonWriter(f), &self.0).map_err(|_| fmt::Error)
```

Would we be able to do something like this? Because I agree with what you said. It's more likely that it's just going to be used to fill some JavaScript variable. However, I could definitely see the case of a debug build, where you're trying to inspect the output

Reply:
Yes, I like this solution. Or could it be confusion this the user that the debug and production built produce different data?

Alternatively, maybe we could add a `|pretty_json` filter?

Reply:
Overall, maybe there should be a `pretty_json`/`tojson_pretty` indeed.

For debugging it would however still require manually changing `tojson` -> `tojson_pretty`. But yeah, explicitness probably wins, compared to the potential confusion.

_Somewhat related, I also think `#[template(..., print = "ast")]` should print using `{:#?}` instead of `{:?}`_